### PR TITLE
layers: Fix ValidateShaderStorageImageFormats: SubpassData

### PR DESCRIPTION
**Pull request by timvpGoogle**
The SPIR-V spec states:

> If the Image Dim operand is not SubpassData, the Image Format must
> not be Unknown, unless the StorageImageReadWithoutFormat Capability was
> declared.

This check is missing, causing ANGLE tests to fail.

### CODE REVIEW

**Comment by sfricke-samsung** [comment]
cc @djdeath as not sure if this needs to anything else with the spec language being touched in https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/4723

File: layers/shader_validation.cpp [hunk 1]
```diff
@@ -1022,7 +1022,10 @@ bool CoreChecks::ValidateShaderStorageImageFormats(SHADER_MODULE_STATE const *sr
             case spv::OpImageRead: {
                 spirv_inst_iter type_def = src->GetImageFormatInst(insn.word(3));
                 if (type_def != src->end()) {
-                    if (type_def.word(8) == spv::ImageFormatUnknown) {
+                    auto dim = type_def.word(3);
+                    // If the Image Dim operand is not SubpassData, the Image Format must not be Unknown, unless the
+                    // StorageImageReadWithoutFormat Capability was declared.
+                    if (dim != spv::DimSubpassData && type_def.word(8) == spv::ImageFormatUnknown) {
```
Comment (djdeath) [inline layers/shader_validation.cpp:1028]: Can this actually happen? Because this is only checking for OpImage(Read|ReadSparse|Write).

Comment (djdeath) [reply layers/shader_validation.cpp:1028]: Oop my bad yep this can happen for read on input attachment.

**Comment by djdeath** [comment]
Thanks for the cc @sfricke-samsung, this looks correct to me. No need to update https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/4723 afaict.

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -1022,7 +1022,10 @@ bool CoreChecks::ValidateShaderStorageImageFormats(SHADER_MODULE_STATE const *sr
             case spv::OpImageRead: {
                 spirv_inst_iter type_def = src->GetImageFormatInst(insn.word(3));
                 if (type_def != src->end()) {
-                    if (type_def.word(8) == spv::ImageFormatUnknown) {
+                    auto dim = type_def.word(3);
```
Comment (ncesario-lunarg) [inline layers/shader_validation.cpp:1025]: `const` (ditto below)?

Comment (timvpGoogle) [reply layers/shader_validation.cpp]: Done

**Comment by timvpGoogle** [comment]
> LGTM! Do we have a test that exercises this?

That's a bit outside of what I'm familiar with.

ANGLE itself is failing several tests without this fix, such as:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/angle/src/tests/gl_tests/MultisampledRenderToTextureTest.cpp;drc=9ddaa68604ae0563743909fa1d27b56a47777fee;l=3871

**Comment by ncesario-lunarg** [comment]
I'm sure someone more apt in the art of crafting raw SPIR-V can create a better, more concise test, but feel free to use this if it looks appropriate. However, as you can see from the `SetUnexpectedError` calls in the test, I'm not sure how to produce a valid SPIR-V binary that hits this particular case without hitting other VUIDs first.
```diff
diff --git a/tests/vklayertests_pipeline_shader.cpp b/tests/vklayertests_pipeline_shader.cpp
index 6c7f9460..a7e79658 100644
--- a/tests/vklayertests_pipeline_shader.cpp
+++ b/tests/vklayertests_pipeline_shader.cpp
@@ -28,6 +28,7 @@
 
 #include "cast_utils.h"
 #include "layer_validation_tests.h"
+#include "core_validation_error_enums.h"
 
 TEST_F(VkLayerTest, PSOPolygonModeInvalid) {
     TEST_DESCRIPTION("Attempt to use invalid polygon fill modes.");
@@ -11482,3 +11483,110 @@ TEST_F(VkLayerTest, ValidateTessellationShaderEnabled) {
                             "VUID-VkShaderModuleCreateInfo-pCode-01091",
                             "VUID-VkPipelineInputAssemblyStateCreateInfo-topology-00430"});
 }
+
+TEST_F(VkLayerTest, SubpassInputWithoutFormat) {
+    TEST_DESCRIPTION("Non-InputAttachment shader input with unknown image format");
+
+    // It appears it would be difficult to hit this particular case without also hitting other VUIDs first
+    m_errorMonitor->SetUnexpectedError("VUID-VkShaderModuleCreateInfo-pCode-01091");
+    m_errorMonitor->SetUnexpectedError(kVUID_Features_shaderStorageImageReadWithoutFormat_NonReadable);
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, kVUID_Features_shaderStorageImageReadWithoutFormat);
+
+    SetTargetApiVersion(VK_API_VERSION_1_2);
+    ASSERT_NO_FATAL_FAILURE(InitFramework());
+
+    if (DeviceValidationVersion() < VK_API_VERSION_1_2) {
+        printf("%s test requires Vulkan 1.2+, skipping test\n", kSkipPrefix);
+        return;
+    }
+
+    VkPhysicalDeviceFeatures features;
+    vk::GetPhysicalDeviceFeatures(gpu(), &features);
+    features.shaderStorageImageReadWithoutFormat = VK_FALSE;
+
+    ASSERT_NO_FATAL_FAILURE(InitState(&features));
+
+    const std::string fsSource = R"(
+               OpCapability Shader
+               OpCapability StorageImageReadWithoutFormat
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %main "main" %color %img
+               OpExecutionMode %main OriginUpperLeft
+               OpSource GLSL 460
+               OpName %main "main"
+               OpName %color "color"
+               OpName %img "img"
+               OpDecorate %color Location 0
+               OpDecorate %img DescriptorSet 0
+               OpDecorate %img Binding 0
+               OpDecorate %img NonWritable
+       %void = OpTypeVoid
+          %3 = OpTypeFunction %void
+      %float = OpTypeFloat 32
+    %v4float = OpTypeVector %float 4
+%_ptr_Output_v4float = OpTypePointer Output %v4float
+      %color = OpVariable %_ptr_Output_v4float Output
+               ;
+               ; Image has unknown format, but dimension != SubpassData and
+               ; shaderStorageImageReadWithoutFormat == VK_FALSE, which is invalid
+               ;
+         %10 = OpTypeImage %float 2D 0 0 0 2 Unknown
+
+%_ptr_UniformConstant_10 = OpTypePointer UniformConstant %10
+        %img = OpVariable %_ptr_UniformConstant_10 UniformConstant
+        %int = OpTypeInt 32 1
+      %v2int = OpTypeVector %int 2
+      %int_0 = OpConstant %int 0
+         %17 = OpConstantComposite %v2int %int_0 %int_0
+       %main = OpFunction %void None %3
+          %5 = OpLabel
+         %13 = OpLoad %10 %img
+         %18 = OpImageRead %v4float %13 %17
+               OpStore %color %18
+               OpReturn
+               OpFunctionEnd
+    )";
+
+    VkShaderObj vs(m_device, bindStateVertShaderText, VK_SHADER_STAGE_VERTEX_BIT, this);
+    VkShaderObj fs(m_device, fsSource, VK_SHADER_STAGE_FRAGMENT_BIT, this, "main", nullptr, SPV_ENV_UNIVERSAL_1_5);
+
+    VkPipelineObj pipe(m_device);
+    pipe.AddShader(&vs);
+    pipe.AddShader(&fs);
+    pipe.AddDefaultColorAttachment();
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    VkDescriptorSetLayoutBinding dslb = {0, VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr};
+    const VkDescriptorSetLayoutObj dsl(m_device, {dslb});
+    const VkPipelineLayoutObj pl(m_device, {&dsl});
+
+    VkAttachmentDescription descs[2] = {
+        {0, VK_FORMAT_R8G8B8A8_UNORM, VK_SAMPLE_COUNT_1_BIT, VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE,
+         VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+         VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL},
+        {0, VK_FORMAT_R8G8B8A8_UNORM, VK_SAMPLE_COUNT_1_BIT, VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE,
+         VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE, VK_IMAGE_LAYOUT_GENERAL, VK_IMAGE_LAYOUT_GENERAL},
+    };
+    VkAttachmentReference color = {
+        0,
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+    };
+    VkAttachmentReference input = {
+        1,
+        VK_IMAGE_LAYOUT_GENERAL,
+    };
+
+    VkSubpassDescription sd = {0, VK_PIPELINE_BIND_POINT_GRAPHICS, 1, &input, 1, &color, nullptr, nullptr, 0, nullptr};
+
+    VkRenderPassCreateInfo rpci = {VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO, nullptr, 0, 2, descs, 1, &sd, 0, nullptr};
+    VkRenderPass rp;
+    VkResult err = vk::CreateRenderPass(m_device->device(), &rpci, nullptr, &rp);
+    ASSERT_VK_SUCCESS(err);
+
+    pipe.CreateVKPipeline(pl.handle(), rp);
+
+    m_errorMonitor->VerifyFound();
+
+    vk::DestroyRenderPass(m_device->device(), rp, nullptr);
+}
```

**Comment by timvpGoogle** [comment]
This bug is blocking rolling VVL into ANGLE and Chromium.   Can we land this and open an issue to add a test later?

**Comment by timvpGoogle** [comment]
> I'm sure someone more apt in the art of crafting raw SPIR-V can create a better, more concise test, but feel free to use this if it looks appropriate. However, as you can see from the `SetUnexpectedError` calls in the test, I'm not sure how to produce a valid SPIR-V binary that hits this particular case without hitting other VUIDs first.
> 
> ```diff
> diff --git a/tests/vklayertests_pipeline_shader.cpp b/tests/vklayertests_pipeline_shader.cpp
> index 6c7f9460..a7e79658 100644
> --- a/tests/vklayertests_pipeline_shader.cpp
> +++ b/tests/vklayertests_pipeline_shader.cpp
> @@ -28,6 +28,7 @@
>  
>  #include "cast_utils.h"
>  #include "layer_validation_tests.h"
> +#include "core_validation_error_enums.h"
>  
>  TEST_F(VkLayerTest, PSOPolygonModeInvalid) {
>      TEST_DESCRIPTION("Attempt to use invalid polygon fill modes.");
> @@ -11482,3 +11483,110 @@ TEST_F(VkLayerTest, ValidateTessellationShaderEnabled) {
>                              "VUID-VkShaderModuleCreateInfo-pCode-01091",
>                              "VUID-VkPipelineInputAssemblyStateCreateInfo-topology-00430"});
>  }
> +
> +TEST_F(VkLayerTest, SubpassInputWithoutFormat) {
> +    TEST_DESCRIPTION("Non-InputAttachment shader input with unknown image format");
> +
> +    // It appears it would be difficult to hit this particular case without also hitting other VUIDs first
> +    m_errorMonitor->SetUnexpectedError("VUID-VkShaderModuleCreateInfo-pCode-01091");
> +    m_errorMonitor->SetUnexpectedError(kVUID_Features_shaderStorageImageReadWithoutFormat_NonReadable);
> +    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, kVUID_Features_shaderStorageImageReadWithoutFormat);
> +
> +    SetTargetApiVersion(VK_API_VERSION_1_2);
> +    ASSERT_NO_FATAL_FAILURE(InitFramework());
> +
> +    if (DeviceValidationVersion() < VK_API_VERSION_1_2) {
> +        printf("%s test requires Vulkan 1.2+, skipping test\n", kSkipPrefix);
> +        return;
> +    }
> +
> +    VkPhysicalDeviceFeatures features;
> +    vk::GetPhysicalDeviceFeatures(gpu(), &features);
> +    features.shaderStorageImageReadWithoutFormat = VK_FALSE;
> +
> +    ASSERT_NO_FATAL_FAILURE(InitState(&features));
> +
> +    const std::string fsSource = R"(
> +               OpCapability Shader
> +               OpCapability StorageImageReadWithoutFormat
> +          %1 = OpExtInstImport "GLSL.std.450"
> +               OpMemoryModel Logical GLSL450
> +               OpEntryPoint Fragment %main "main" %color %img
> +               OpExecutionMode %main OriginUpperLeft
> +               OpSource GLSL 460
> +               OpName %main "main"
> +               OpName %color "color"
> +               OpName %img "img"
> +               OpDecorate %color Location 0
> +               OpDecorate %img DescriptorSet 0
> +               OpDecorate %img Binding 0
> +               OpDecorate %img NonWritable
> +       %void = OpTypeVoid
> +          %3 = OpTypeFunction %void
> +      %float = OpTypeFloat 32
> +    %v4float = OpTypeVector %float 4
> +%_ptr_Output_v4float = OpTypePointer Output %v4float
> +      %color = OpVariable %_ptr_Output_v4float Output
> +               ;
> +               ; Image has unknown format, but dimension != SubpassData and
> +               ; shaderStorageImageReadWithoutFormat == VK_FALSE, which is invalid
> +               ;
> +         %10 = OpTypeImage %float 2D 0 0 0 2 Unknown
> +
> +%_ptr_UniformConstant_10 = OpTypePointer UniformConstant %10
> +        %img = OpVariable %_ptr_UniformConstant_10 UniformConstant
> +        %int = OpTypeInt 32 1
> +      %v2int = OpTypeVector %int 2
> +      %int_0 = OpConstant %int 0
> +         %17 = OpConstantComposite %v2int %int_0 %int_0
> +       %main = OpFunction %void None %3
> +          %5 = OpLabel
> +         %13 = OpLoad %10 %img
> +         %18 = OpImageRead %v4float %13 %17
> +               OpStore %color %18
> +               OpReturn
> +               OpFunctionEnd
> +    )";
> +
> +    VkShaderObj vs(m_device, bindStateVertShaderText, VK_SHADER_STAGE_VERTEX_BIT, this);
> +    VkShaderObj fs(m_device, fsSource, VK_SHADER_STAGE_FRAGMENT_BIT, this, "main", nullptr, SPV_ENV_UNIVERSAL_1_5);
> +
> +    VkPipelineObj pipe(m_device);
> +    pipe.AddShader(&vs);
> +    pipe.AddShader(&fs);
> +    pipe.AddDefaultColorAttachment();
> +    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
> +
> +    VkDescriptorSetLayoutBinding dslb = {0, VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr};
> +    const VkDescriptorSetLayoutObj dsl(m_device, {dslb});
> +    const VkPipelineLayoutObj pl(m_device, {&dsl});
> +
> +    VkAttachmentDescription descs[2] = {
> +        {0, VK_FORMAT_R8G8B8A8_UNORM, VK_SAMPLE_COUNT_1_BIT, VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE,
> +         VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
> +         VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL},
> +        {0, VK_FORMAT_R8G8B8A8_UNORM, VK_SAMPLE_COUNT_1_BIT, VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE,
> +         VK_ATTACHMENT_LOAD_OP_LOAD, VK_ATTACHMENT_STORE_OP_STORE, VK_IMAGE_LAYOUT_GENERAL, VK_IMAGE_LAYOUT_GENERAL},
> +    };
> +    VkAttachmentReference color = {
> +        0,
> +        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
> +    };
> +    VkAttachmentReference input = {
> +        1,
> +        VK_IMAGE_LAYOUT_GENERAL,
> +    };
> +
> +    VkSubpassDescription sd = {0, VK_PIPELINE_BIND_POINT_GRAPHICS, 1, &input, 1, &color, nullptr, nullptr, 0, nullptr};
> +
> +    VkRenderPassCreateInfo rpci = {VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO, nullptr, 0, 2, descs, 1, &sd, 0, nullptr};
> +    VkRenderPass rp;
> +    VkResult err = vk::CreateRenderPass(m_device->device(), &rpci, nullptr, &rp);
> +    ASSERT_VK_SUCCESS(err);
> +
> +    pipe.CreateVKPipeline(pl.handle(), rp);
> +
> +    m_errorMonitor->VerifyFound();
> +
> +    vk::DestroyRenderPass(m_device->device(), rp, nullptr);
> +}
> ```

I created a new PR with this test:
https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3070/

I don't know how to actually run it or verify it or anything, but wanted to at least get the CL out there for you.

**Comment by ncesario-lunarg** [comment]
Thanks @timvpGoogle!

@sfricke-samsung @jeremy-lunarg do either of you have any thoughts on what would be a good test case here?

**Pull request merged**